File: src/morie/fn/sgtlap.py
# morie.fn -- function file (rootcoder007/morie)
"""Combinatorial graph Laplacian L = T - A."""

from . import _tail1core as C

from ._richresult import RichResult

__all__ = ["graphlap", "sgt_laplacian"]


def graphlap(W):
    """Combinatorial (unnormalised) Laplacian of a weighted graph.

    ``rowsum`` is returned because it must be zero to machine
    precision: the all-ones vector is always in the kernel, and a
    non-zero row sum means the weight matrix was not symmetric or the
    loop convention was mishandled.

    Formula: L(u, v) = d_v - w(v, v) if u = v, -w(u, v) if u ~ v, else 0

    Parameters
    ----------
    W : array-like, shape (n, n)
        Symmetric non-negative weight matrix.

    Returns
    -------
    RichResult
        ``L``, ``degree``, ``rowsum``, ``n``.

    References
    ----------
    Chung (1997), Spectral Graph Theory, CBMS 92, Section 1.4, which
    generalises the Section 1.2 definition to weighted graphs with
    loops exactly as written above.  Fetched from the author's own
    copy of the chapter.
    """
    W = C.mat(W)
    n = len(W)
    if any(len(r) != n for r in W):
        raise ValueError("W must be square")
    d = [sum(W[i]) for i in range(n)]
    L = [[(d[i] - W[i][i]) if i == j else -W[i][j] for j in range(n)]
         for i in range(n)]
    return RichResult(payload={
        "L": L, "degree": d, "rowsum": [sum(r) for r in L], "n": n,
        "method": "Combinatorial Laplacian L = T - A"})


sgt_laplacian = graphlap


def cheatsheet():
    return "sgtlap: L(u,v) = d_v - w(v,v) on the diagonal, -w(u,v) off it"

File: r-package/morie/R/sgtlap.R
# SPDX-License-Identifier: AGPL-3.0-or-later
#' Combinatorial graph Laplacian.
#'
#' Formula: L = D - A
#'
#' @param Adj Symmetric adjacency matrix with a zero diagonal; entries may be 0/1 or non-negative edge weights.
#'
#' @return List with ``L``, ``degrees``, ``n``.
#' @references Chung, F. (1997), Spectral Graph Theory, CBMS Regional Conference Series in Mathematics 92, American Mathematical Society, is this shelf's primary book and is NOT in the reference library, so it could not be read.  The conventions below were taken instead from the author's own survey, Chung, F., Four proofs for the Cheeger inequality and graph partition algorithms, Proceedings of ICCM 2007 Vol. II pp. 1-4, Sect. 2 (Preliminaries), which restates them in her notation; that paper was FETCHED and is archived in the reference library with a row in EXTERNAL_SOURCES.md.  This particular matrix is not printed in the retrieved survey and is implemented in the standard published form; it should be re-checked against Chung (1997) Chapter 1 if the book is ever added to the library.  The combinatorial Laplacian is positive semi-definite with the all-ones vector in its kernel, so its smallest eigenvalue is always zero.
#' @export
Glaplac <- function(Adj) {
  A <- .t1_mat(Adj); n <- nrow(A)
  if (n == 0L || ncol(A) != n) stop("Adj must be a non-empty square matrix")
  if (any(A < 0)) stop("edge weights must be non-negative")
  if (any(A != t(A))) stop("Adj must be symmetric")
  d <- rowSums(A)

  .t1_result(L = diag(d, n) - A, degrees = d, n = n,
             method = "Combinatorial Laplacian D - A")
}
